ls1intum/Artemis
# `General`: Provide overview over course lifetime in active students chart on course detail

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


The active students' chart on course detail currently provides an immutable scope size of 17 weeks. Given that the course as a start date defined, this PR implements a new scope: the course lifetime overview. This scope contains all weeks starting from the course start date up to the current date.
The current scope (17 weeks) is displayed per default for active courses. The user can switch between the two scopes if desired.
If the course is already finished, the lifetime overview is shown per default. Again, the scope can be switched by the user.

What else has been done:

- I modified the date validation mechanism for courses. Currently, it was not possible to delete end dates or start dates once set and saved.
- I removed the "CV"-prefix from the x-axis ticks of the chart as they are redundant. Instead, the chart now provides a meaningful x-axis label that is translation sensitive.

### Description

I added the corresponding selection functionality via a switch in the UI.
To get the corresponding data, I created a new endpoint that sends this data to the client.
I implemented the client logic this way that the data for both initial scopes are fetched at most once. As the data should not change this often, this prevents too many server and database calls.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with student activity and no start or end date

1. Log in to Artemis
2. Navigate to Course Administration
3. Enter the course detail page 
4. Inspect the chart. It should look as usual.
5. Edit the course and set a course start date.
6. Return to the detail page. Now, a switch should be displayed. "Recent activity" should be selected per default.
7. Check the tooltips for the different parts of the switch. Also, check the german translation. In this context, confirm that the x-axis label instantly translates after changing the language settings.
8. Change the scope to "Lifetime overview". Inspect the chart and confirm that everything looks as expected
9. Edit the course again. Set a course end date in the past.
10. Return to the detail page. Now, "Lifetime overview" should be selected per default. Switch to the limited scope again and confirm that everything works.
11. Edit the page again and try to delete course start and end date. It should be possible to save this configuration now.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-detail-line-chart.component.ts | 80% | 95% |
| course-management.service.ts | 81.48% | 91.66% |
| CourseService.java | 76% | 90% |
| CourseResource.java | 73% | 93% |


### Screenshots


Active course with start date:
![image](https://user-images.githubusercontent.com/80622272/171616674-c38231b3-70ec-4055-a68f-c7e4b79053a5.png)

Lifetime overview scope:
![image](https://user-images.githubusercontent.com/80622272/171616784-099d3f77-874a-4a70-977b-5bc0b7c38dbf.png)
